app/routes/_auth+/signup.tsx

```diff
@@ -25,18 +25,56 @@ export const onboardingEmailSessionKey = 'onboardingToken'
 const onboardingTokenQueryParam = 'token'
 const tokenType = 'onboarding'
 
-const signupSchema = z.object({
-	email: emailSchema.refine(
-		async email => {
-			const existingUser = await prisma.user.findUnique({
-				where: { email },
-				select: { id: true },
+function createSchema(
+	intent: string = "",
+	constraints: {
+		isEmailUnique: (email: string) => Promise<boolean>;
+	} = {
+		isEmailUnique: async () => true,
+	},
+) {
+	const signupSchema = z.object({
+		email: emailSchema
+			.superRefine((email, ctx) => {
+				if (intent !== 'validate/email' && intent !== 'submit') {
+					// Validate only when the email field is changed or when submitting
+					ctx.addIssue({
+						code: z.ZodIssueCode.custom,
+						message: conform.VALIDATION_SKIPPED,
+					});
+				} else if (typeof constraints.isEmailUnique === 'undefined') {
+					// Validate only if the constraint is defined
+					ctx.addIssue({
+						code: z.ZodIssueCode.custom,
+						message: conform.VALIDATION_UNDEFINED,
+					});
+				} else {
+					// Tell zod this is an async validation by returning the promise
+					return constraints.isEmailUnique(email).then((isUnique) => {
+						if (isUnique) {
+							return;
+						}
+
+						ctx.addIssue({
+							code: z.ZodIssueCode.custom,
+							message: 'Email is already used',
+						});
+					});
+				}
 			})
-			return !existingUser
-		},
-		{ message: 'A user already exists with this email' },
-	),
-})
+			.refine(
+				async email => {
+					const existingUser = await prisma.user.findUnique({
+						where: { email },
+						select: { id: true },
+					})
+					return !existingUser
+				},
+				{ message: 'A user already exists with this email' },
+			),
```

Comment:
This should be removed with the previous `.superRefine()`

Reply:
Done. Thanks for the feedback. Simpler now